ESQL: Fix enrich and lookup join resolution based on min transport version

alex-spies opened the pull request:
Relates https://github.com/issues/created?issue=elastic%7Celasticsearch%7C137534

When deciding which types are supported, we did not use the correct minimum transport version during the enrich resolution step. What's more, the `EnrichPolicyResolver` did not account for the fact that the node requesting resolution might be on a version that doesn't support the types in the resolved mapping, which led to serialization bugs surfacing when trying to enable the `DATE_RANGE` type. We had inconsistent minimum transport version in case of `ROW | ENRICH`/`ROW | LOOKUP JOIN` queries and `FROM remote:* | ENRICH` queries.

For `ENRICH`, this is a non-issue in practice, because the only affected data types were `AGGREGATE_METRIC_DOUBLE` and `DENSE_VECTOR`, which cannot yet be used in enrich policies to begin with. (See https://github.com/elastic/elasticsearch/issues/137699 and https://github.com/elastic/elasticsearch/issues/127350).

`ROW | LOOKUP JOIN` queries were broken (or at least potentially inconsistent) because a new coordinator might have assumed that the two new data types are supported, while an old lookup node doesn't actually support them.

- [x] Pass the determined overall minimum transport version into the query profile for debugging and testing.
- [x] When resolving enrich policies, pass the minimum version determined in the main index resolution to the enrich policy resolver. This is important for CCS, where the enrich policies are resolved locally on the remote clusters.
- [x] For consistency, do the same for lookup joins (technically not required, they perform their field caps call on the coordinator which is simpler).
- [x] ~For both, enrich and lookup resolution, retrieve the determined minimum version and use it on the coordinator. This is important for `ROW | ENRICH` and `ROW | LOOKUP JOIN` queries.~
- [x] Retrieve the coordinating cluster's minimum transport version from the cluster state and use that as baseline so that `ROW | ENRICH` and `ROW | LOOKUP JOIN` queries don't just use the coordinator node's version (which may be newer than the version of a node executing enrich/lookup joins for us).
- [x] Add a bunch of tests
- [x] Add tests for `FROM remote_only:*`
- [x] Add tests for `FROM | ENRICH`, both for queries on the local cluster and in case of CCS.
- [x] Add tests for `FROM | LOOKUP JOIN` on the local cluster + with CCS.

alex-spies committed:
Restrict EnrichPolicyResolver to supported types

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Update docs/changelog/137431.yaml

alex-spies committed:
Plumb minimum version into ENRICH resolution

Commit:
[CI] Update transport version definitions

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-enrich-lj-resolution-min-transport-version

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

Commit:
[CI] Update transport version definitions

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -84,58 +84,88 @@ public IndexResolver(Client client) {
         boolean includeAllDimensions,
         boolean useAggregateMetricDoubleWhenNotSupported,
         boolean useDenseVectorWhenNotSupported,
+        @Nullable TransportVersion minimumVersion,
```
This `@Nullable` is a bit confusing. This method is private to the class and has only two usages:
* one in `resolveAsMergedMappingForVersion` that never supplies null
* one in `resolveAsMergedMapping` that allways supplies null

Lets change the second one to supply min compatible to avoid nullability and a condition below.

alex-spies replied:
Had to re-do it because I was wrong initially :D there's only 1 method now, and it has a nullable minimum version to be passed in, intentionally.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -84,58 +84,88 @@ public IndexResolver(Client client) {
         QueryBuilder requestFilter,
         boolean includeAllDimensions,
         boolean useAggregateMetricDoubleWhenNotSupported,
         boolean useDenseVectorWhenNotSupported,
```
Since you are changing the signature, those 3 booleans are always false when supplied to this method. The signature could be simplified.

alex-spies replied:
I removed this method again to simplify.

I considered whether the `use...WhenNotSupported` booleans can be removed, but they're still required in rolling upgrades when one of the nodes is 9.2.0. This is the only version that's already supporting `aggregate_metric_double` and `dense_vector`, but doesn't provide its transport version in field caps responses.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -610,14 +615,15 @@ private void preAnalyzeLookupIndex(
             false,
             false,
+            result.minimumTransportVersion(),
             listener.map(indexResolution -> receiveLookupIndexResolution(result, localPattern, executionInfo, indexResolution))
```
I think I still made a mistake in the current state of the PR. I think, the transport version of all the requests (main resolution, lookup resolution, enrich policy resolution) needs to be taken into account.

Gonna fix + add tests. That said:

1. `ROW | LOOKUP JOIN`: the main index resolution doesn't fire a field cap request, and thus we assume 
`TransportVersion.current()` to be the minimum version overall. Which is okay for the `ROW`, but the `LOOKUP JOIN` will be executed on another node, which may be behind.
2. `FROM remote*:idx | ENRICH`: We send remote requests to the `EnrichPolicyResolver`s of the remotes, and those perform a cluster-local field caps request. I believe they won't take into account that the cluster that fired the request may be behind, and thus we may send them back an index mapping with types that the calling cluster can't even deserialize - boom.

alex-spies replied:
> > ROW | LOOKUP JOIN

> I believe is fully executed on a local coordinator.

In general, no, because the coordinator may not have a replica of the lookup index. Even if it does, I'm not sure if all lookup requests always go to the local coordinator - we perform a new request per page of data to be looked up.

> > FROM remote*:idx | ENRICH

> Do we send such to all remotes? I am a bit surprised and expected this to be narrower.

I think we do, take a look at where we send the enrich resolution request to the remotes [in the policy resolver](https://github.com/elastic/elasticsearch/blob/4fb85c605839cadf0f0bdcb95834cb9d296f5e72/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java#L318-L327).

(Seems to still happen when we specifically make the ENRICH coordinator-only, because our pre-analysis [doesn't seem to care](https://github.com/elastic/elasticsearch/blob/4fb85c605839cadf0f0bdcb95834cb9d296f5e72/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/PreAnalyzer.java#L64).)

alex-spies committed:
Plumb minimumVersion into the profile

alex-spies committed:
Update tests

alex-spies committed:
Assert minimum version in AllSupportedFields tests

alex-spies committed:
Fix tests some more

alex-spies committed:
Fix test

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-enrich-lj-resolution-min-transport-version

alex-spies committed:
Fix test expectations

Commit:
[CI] Update transport version definitions

alex-spies committed:
Add test for ROW

alex-spies committed:
Remove obsolete expectations

DENSE_VECTOR is completely unsupported on 9.1, no need to check for this
version and adjust expectations.

alex-spies committed:
WIP: LOOKUP JOIN tests

alex-spies committed:
Consider min version from LOOKUP JOIN resolution

Important for ROW queries.

alex-spies committed:
Fix serialization exception

alex-spies committed:
Finish test for cluster-local LOOKUP JOIN

alex-spies committed:
Rename minimumVersion->minimumTransportVersion in profile

alex-spies committed:
Add test with ROW | ENRICH

alex-spies committed:
Fix + test ROW | ENRICH

alex-spies committed:
Delete docs/changelog/137431.yaml

alex-spies committed:
Update docs/changelog/137431.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Delete docs/changelog/137431.yaml

alex-spies committed:
Update docs/changelog/137431.yaml

alex-spies committed:
Fix bug

alex-spies committed:
Add comment

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-enrich-lj-resolution-min-transport-version

Commit:
[CI] Update transport version definitions

alex-spies committed:
Remove leftovers

alex-spies committed:
Clarify comment

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -84,58 +84,58 @@ public IndexResolver(Client client) {
-                LOGGER.debug("minimum transport version {} {}", response.caps().minTransportVersion(), info.effectiveMinTransportVersion());
-                l.onResponse(new Versioned<>(mergedMappings(indexWildcard, info), info.effectiveMinTransportVersion()));
+                LOGGER.debug(
+                    "updated minimum transport version from {} to effective version {} using version {} from field caps response",
```
```suggestion
                    "updated minimum transport version from [{}] to effective version [{}] using version [{}] from field caps response",
```

alex-spies committed:
Improve log message

alex-spies committed:
Rename resolveAsMergedMapping->resolveIndexPattern

alex-spies committed:
Obtain local cluster version from ClusterState

Do not propagate the minimum transport version determined during ENRICH
and LOOKUP JOIN resolution back to the resolution pipeline in
EsqlSEssion. That's overkill just for ROW queries. We can get the
minimum version of the cluster simply from the ClusterState.

alex-spies committed:
Update comment

alex-spies committed:
Fix comment

alex-spies committed:
Fix some more comments

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-enrich-lj-resolution-min-transport-version

alex-spies committed:
Minor simplification

alex-spies committed:
Fix tests

alex-spies committed:
Fix comment

alex-spies committed:
Simplify leftovers

Commit:
[CI] Update transport version definitions

alex-spies committed:
Fix checkstyle violations

alex-spies committed:
Handle null values in profile's minimum version

alex-spies committed:
Fix another test

alex-spies committed:
Add test for FROM *:* (only remote indices)

alex-spies committed:
Add test for FROM | ENRICH

And fix expectations for the determined min cluster version in non-CCS
tests.

alex-spies committed:
Add test for FROM | LOOKUP JOIN

alex-spies committed:
Force lookup joins onto remotes for good measure

alex-spies committed:
Make assertion stricter in test

craigtaverner committed:
Merge branch 'main' into fix-enrich-lj-resolution-min-transport-version

alex-spies committed:
Merge branch 'main' into fix-enrich-lj-resolution-min-transport-version

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-enrich-lj-resolution-min-transport-version

Commit:
[CI] Update transport version definitions

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  9.2  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 137431`
